Add context to inventory plugins documentation

JonTheNiceGuy opened the pull request:
##### SUMMARY
* Show how to add the inventory plugin to the `enable_plugins` configuration item - not just replace it.
* Show adding tags to `keyed_groups` without specifying the separator.
* Show adding more complex `groups` and `compose` statements, including using the ternary filter.
* As I added an additional group, I've updated the resulting `ansible-inventory` output accordingly.

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
ansible.builtin.constructed_inventory

##### ADDITIONAL INFORMATION
This has been raised in response to confusion I had, which was cleared up during conversations with @s-hertel and @flowerysong in the Ansible-AWS IRC channel.

JonTheNiceGuy committed:
Add context to inventory plugins documentation

* Show how to add the inventory plugin to the `enable_plugins` configuration item - not just replace it.
* Show adding tags to `keyed_groups` without specifying the separator.
* Show adding more complex `groups` and `compose` statements, including using the ternary filter.
* As I added an additional group, I've updated the resulting `ansible-inventory` output accordingly.

s-hertel commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -80,12 +86,24 @@ You can create dynamic groups using host variables with the constructed ``keyed_
       - key: tags.Name
         prefix: tag_Name_
         separator: ""
+      # If you have a tag called "Role" which has the value "Webserver", this will add the group
```
This is demonstrating the same thing as the tags.Name example, so maybe we can improve the wording there.

JonTheNiceGuy replied:
So, `tags.Name` provides an indicator of the role that the "separator" flag provides. Perhaps, instead, we should put something different in there - perhaps using some other tag like this:

```yml
- key: tags.Features
  prefix: Feature
  separator: "%"
```

And show the groups:

```
|--@Feature%Blue
|  |--abc123
|--@Feature%Green
|  |--def456
```

If this is an acceptable counter-offer, I'll make this change instead.

s-hertel replied:
Oh, true. We shouldn't recommend using special characters in group names though. The updated examples look good to me, but it was pointed out to me that this page should try to be a simple overview, and these should probably go to the plugin-specific docs for constructed here https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/constructed.py#L26. You could add a comment to this page though, to see the constructed plugin for more examples if there isn't already one.

We might be able to clarify the wording for this comment - initially I read it as create groups for the tag "Role" if it has the value "Webserver", whereas the role_value groups are formed regardless of the value. But I'll leave that to the docs team.

JonTheNiceGuy replied:
OK, so how about this, we yank all of the separator stuff out of here, so it just becomes:

```
- key: tags.Role
  prefix: role
```

(which, I think is generally a better and more accurate example than "this host was named something" as you don't tend to have lots of machines with the same name...)

Then, we fix up the constructed document with some of the examples from here (removing them from this as a result), perhaps including:

```
compose:
  ansible_host: public_ip_address is defined | ternary(public_ip_address, private_ip_address)
  some_inventory_wide_string: '"Yes, you need both types of quotes here"'
```

Akasurde replied:
@s-hertel Could you please help here? Thanks in advance.

s-hertel commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -80,12 +86,24 @@ You can create dynamic groups using host variables with the constructed ``keyed_
       # add hosts to the group development if any of the dictionary's keys or values is the word 'devel'
       development: "'devel' in (tags|list)"
+      # add hosts to the "private_only" group if the host doesn't have a public IP associated to it
+      private_only: "(network_interfaces[0].association | default({}) ).public_ip | default('') != ''"
```
As long as public_ip isn't sometimes an empty string, this could be
```suggestion
      private_only: "(network_interfaces[0].association | default({}) ).public_ip is defined"
```

Why not check the top level variable `public_ip_address` though?

JonTheNiceGuy replied:
Ah, so let's cover off the "why not check the top level variable" - because I completely missed it! :) I looked for it on a node that didn't have a public IP, and it wasn't there, so I figured I'd have to "find it" somewhere else, and then went looking for where the attachment happened elsewhere - `network_interfaces`. Having now looked for it, I can clearly see it. D'oh.

Also yes, `is defined` versus `| default('') != ''` makes much more sense. I have hit issues with this in the dim and distant past, mostly with parsing environment variables, and so my "default reaction" is to go with the latter form rather than the former. Yes, I need to fix this one!

JonTheNiceGuy replied:
Oh, an interesting quirk. `public_ip_address` is only defined where the first interface has a public ip defined. If you attach an EIP to the second or later interfaces, then it's not classed as that host's public ip address. This is totally outside the scope of this documentation change, but it gives me a bit more context for why I might have missed this before :)

s-hertel commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -80,12 +86,24 @@ You can create dynamic groups using host variables with the constructed ``keyed_
-      # set the ansible_host variable to connect with the private IP address without changing the hostname
-      ansible_host: private_ip_address
+      # use a private address where a public one isn't assigned
+      ansible_host: ((network_interfaces[0].association | default({}) ).public_ip | default('') != '') | ternary((network_interfaces[0].association | default({}) ).public_ip, private_ip_address)
```
You don't need to use `|default({})` in the ternary filter since the public ip was found.

JonTheNiceGuy replied:
Of course. Good spot.

s-hertel reviewed (commented):
Thanks for sharing your findings back

bcoca commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -31,6 +31,12 @@ If the plugin is in a collection, use the fully qualified name:
+.. code-block:: ini
+
+   [inventory]
+   enable_plugins = host_list, script, auto, yaml, ini, toml, namespace.collection_name.inventory_plugin_name
```
i would note that if plugin author uses the `plugin: fqcn` convention this is not needed as the `auto` plugin should just load the appropriate plugin

JonTheNiceGuy replied:
OK, I'll note that, instead of the convoulted route I took, you should instead make sure you're using the fqcn.

JonTheNiceGuy committed:
Update inventory.rst

Resolved issue addressed in https://github.com/ansible/ansible/pull/72292#discussion_r510192549

JonTheNiceGuy committed:
Update inventory.rst

Resolved https://github.com/ansible/ansible/pull/72292#discussion_r510193507 by removing the whole stanza in question, and replacing it with the more sensible one addressed in https://github.com/ansible/ansible/pull/72292#discussion_r510192549

JonTheNiceGuy committed:
Update inventory.rst

Tidying up the mangled content spotted in https://github.com/ansible/ansible/pull/72292/files/7de71bcb5cbfbcc15ca32f0bf743dac5e85f19fd#r510198484

JonTheNiceGuy commented:
Right, @bcoca, @s-hertel, I've made most of the changes suggested. Can you review this again please? The only outstanding one is the `keyed_groups` with separator versus without noted by @s-hertel.

s-hertel commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -24,13 +24,19 @@ In some circumstances, for example, if the inventory plugin does not use a YAML
-   enable_plugins = namespace.collection_name.inventory_plugin_name
+   enable_plugins = host_list, script, auto, yaml, ini, toml, namespace.collection_name.inventory_plugin_name
 
+Or, if it's a local plugin, perhaps stored in `plugins/inventory/my_plugin.py`, you could reference it like this:
```
Maybe use the default path for local plugins instead https://docs.ansible.com/ansible/latest/dev_guide/developing_locally.html.

JonTheNiceGuy replied:
Resolved in 8511d14

s-hertel commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -80,12 +86,24 @@ You can create dynamic groups using host variables with the constructed ``keyed_
         prefix: tag_Name_
         separator: ""
+      # If you have a tag called "Role" which has the value "Webserver", this will add the group
+      # Role_Webserver and add any hosts that have that tag assigned to it.
```
role_Webserver

JonTheNiceGuy replied:
Fixed in 6a55691

JonTheNiceGuy committed:
Update inventory.rst

Using a better path as suggested in https://github.com/ansible/ansible/pull/72292/files/93781fa3e6900fcaac4f6bea6fe6fedf2a71ae19#r511018431

JonTheNiceGuy committed:
Update inventory.rst

Fix typo, identified in https://github.com/ansible/ansible/pull/72292/files/93781fa3e6900fcaac4f6bea6fe6fedf2a71ae19#r511020771

Akasurde commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -24,13 +24,19 @@ In some circumstances, for example, if the inventory plugin does not use a YAML
    enable_plugins = host_list, script, auto, yaml, ini, toml
 
-If the plugin is in a collection, use the fully qualified name:
+If the plugin is in a collection, and isn't being picked up by the `auto` statement, you can append the fully qualified name:
```
```suggestion
If the plugin is in a collection and is not being picked up by the `auto` statement, you can append the fully qualified name:
```

Akasurde commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -24,13 +24,19 @@ In some circumstances, for example, if the inventory plugin does not use a YAML
-   enable_plugins = namespace.collection_name.inventory_plugin_name
+   enable_plugins = host_list, script, auto, yaml, ini, toml, namespace.collection_name.inventory_plugin_name
 
+Or, if it's a local plugin, perhaps stored in `~/.ansible/plugins/inventory/my_plugin.py`, you could reference it like this:
```
```suggestion
Or, if it is a local plugin, perhaps stored in `~/.ansible/plugins/inventory/my_plugin.py`, you could reference it like this:
```

JonTheNiceGuy committed:
Update docs/docsite/rst/plugins/inventory.rst

Co-authored-by: Abhijeet Kasurde <akasurde@redhat.com>

JonTheNiceGuy committed:
Update docs/docsite/rst/plugins/inventory.rst

Co-authored-by: Abhijeet Kasurde <akasurde@redhat.com>

bcoca commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -31,7 +31,7 @@ If the plugin is in a collection and is not being picked up by the `auto` statem
    enable_plugins = host_list, script, auto, yaml, ini, toml, namespace.collection_name.inventory_plugin_name
 
-Or, if it's a local plugin, perhaps stored in `~/.ansible/plugins/inventory/my_plugin.py`, you could reference it like this:
+Or, if it is a local plugin, perhaps stored in `~/.ansible/plugins/inventory/my_plugin.py`, you could reference it like this:
```
i would not give this path but the 'setting' that cointains this and other paths  DEFAULT_INVENTORY_PLUGIN_PATH (link to it's doc page, which shows the defaults)

samccann commented on docs/docsite/rst/plugins/inventory.rst:
```diff
@@ -24,13 +24,19 @@ In some circumstances, for example, if the inventory plugin does not use a YAML
-   enable_plugins = namespace.collection_name.inventory_plugin_name
+   enable_plugins = host_list, script, auto, yaml, ini, toml, namespace.collection_name.inventory_plugin_name
 
+Or, if it is a local plugin, perhaps stored in `~/.ansible/plugins/inventory/my_plugin.py`, you could reference it like this:
```
```suggestion
Or, if it is a local plugin, perhaps stored in the path set by :ref:`DEFAULT_INVENTORY_PLUGIN_PATH`, you could reference it as follows:
```

samccann commented:
Hi @JonTheNiceGuy ! Sorry once again for the longish delays. I put in a suggested edit to cover @bcoca's comment.

@s-hertel - can you comment or agree with the other proposed changes from https://github.com/ansible/ansible/pull/72292/files#r511119184 ? 

Thanks!

s-hertel commented:
@JonTheNiceGuy /@samccann You could leave it as it is or implement https://github.com/ansible/ansible/pull/72292#discussion_r511119184, I don't have a preference. It's not a blocking comment, for sure. Thanks for improving the docs.

samccann committed:
Update docs/docsite/rst/plugins/inventory.rst
